workflow/validate/validate.go

```diff
@@ -1017,6 +1049,28 @@ func validateOutputParameter(paramRef string, param *wfv1.Parameter) error {
 	return nil
 }
 
+// getNameFieldValue gets the value of the `Name` field from the struct
+// if not found from the struct, it checks embedded field instead
+func getNameFieldValue(s reflect.Value) (string, error) {
```

Comment:
Reflection can make bugs hard to find and fix. You cannot perform refracting when using reflection for example. So you write more tests. Is there any way we can avoid it?

Reply:
Thanks! Totally agree. I think to avoid reflect completely, the most straight-forward way is to just make the validation method accept `[]string` of names instead:

```diff
--- a/workflow/validate/validate.go
+++ b/workflow/validate/validate.go
@@ -1022,7 +1022,7 @@ func validateOutputParameter(paramRef string, param *wfv1.Parameter) error {
 // * unique
 // * non-empty
 // * matches matches our regex requirements
-func validateWorkflowFieldNames(slice interface{}) error {
+func validateWorkflowFieldNames(names []string, isParamOrArtifact bool) error {
```

Note that we need `isParamOrArtifact` because parameters and artifact has different name validation regex than other fields.

When we use it, we create a slice of names instead:

```go
names := []string{}
for _, t := range wf.Spec.Templates {
	names = append(names, t.Name)
}
validateWorkflowFieldNames(names, false)
```

What do we think?

Reply:
I agree